### PR TITLE
NuoDB External access doc

### PR DESCRIPTION
**Pull request by sivanov-nuodb**
https://github.com/nuodb/nuodb-helm-charts/blob/sivanov/DOC-3444/docs/HowToConnectExternally.md

### CODE REVIEW

**Review by acabrele — APPROVED** [review]
Great new feature!

**Review by bkelly-ndb — APPROVED** [review]

File: docs/HowToConnectExternally.md [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+A _direct_ connection to TEs is also supported by NuoDB, however, in this case, the load balancing must be performed by the client application.
+
+SQL clients and applications running in the same Kubernetes cluster as the NuoDB domain can connect to the database with the default NuoDB Helm Charts configuration using the NuoDB Admin `ClusterIP` service or directly using the TE `ClusterIP` service. This document focuses on external client applications running outside of the Kubernetes cluster where the NuoDB database is hosted. Allowing external access to the NuoDB database is not enabled by default and requires additional configuration.
+
+## Transaction Engine Groups
+
+A TE group consists of TEs with the same configuration which is part of the same database most often used to serve specific SQL workload.
+NuoDB Helm Charts 3.4.0+ supports the deployment of one or more TE groups per database.
+
+Multiple deployments of the `database` chart for the same NuoDB database can be done in the same Kubernetes namespace where only one of them is _primary_.
```
[the remark this answers is not in the document]
Comment (adriansuarez) [reply docs/HowToConnectExternally.md:25]: Also, is "deployments" the correct term, or "releases"? I think "deployments" might be overloaded.

File: docs/HowToConnectExternally.md [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+One or more _secondary_ Helm releases are used to deploy additional TE groups for the same database with different configuration options.
+This allows flexible configuration of each TE group including but not limited to the number of TEs in a group, their resource requirements, process labels, and scheduling rules.
```
Comment (adriansuarez) [inline docs/HowToConnectExternally.md:27]: Can we reword this to make it clearer that TEs within the same group have identical process labels (used by LBQuery) and K8s scheduling rules, and can be scaled independently of each other?

File: docs/HowToConnectExternally.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (sivanov-nuodb) [reply docs/HowToConnectExternally.md:25]: Good point! I've updated these.

File: docs/HowToConnectExternally.md [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+In a case of a problem during NLB provisioning, the IP address will never be populated and the _engine_ container will fail.
+The following errors can be seen in the TE container logs:
+
+```text
+2021-12-15T07:43:05.015+0000 INFO  [admin-nuodb-cluster0-0:te-database-nuodb-cluster0-demo-55d664c8d7-bn57p] CustomAdminCommands Found service name=database-nuodb-cluster0-demo-balancer, type=LoadBalancer, selector={u'app': u'database-nuodb-cluster0-demo', u'component': u'te'}
+2021-12-15T07:43:05.015+0000 INFO  [admin-nuodb-cluster0-0:te-database-nuodb-cluster0-demo-55d664c8d7-bn57p] CustomAdminCommands Waiting for load balancer service database-nuodb-cluster0-demo-balancer ingress address...
+'start te' failed: Timeout after 120.0 sec waiting for ingress hostname in service database-nuodb-cluster0-demo-balancer
+```
+
+> **ACTION**: Verify that the `EXTERNAL-IP` for the service is available using `kubectl get services` command.
```
Comment (jleslie85) [inline docs/HowToConnectExternally.md:296]: If using the word command at the end, then I suggest adding "the" after using.  Or simply just remove the trailing word "command" as not needed.

File: docs/HowToConnectExternally.md [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+Check the events for the service for this TE group using `kubectl describe service database-nuodb-cluster0-demo-balancer` command.
```
Comment (jleslie85) [inline docs/HowToConnectExternally.md:297]: Same comment.

File: docs/HowToConnectExternally.md [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+Look into the cloud provider documentation on how to troubleshoot the load balancer controller.
```
Comment (jleslie85) [inline docs/HowToConnectExternally.md:298]: Suggest "Refer to your cloud provider's" instead of "Look into the ..."

File: docs/HowToConnectExternally.md [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+The Kubernetes cluster should be properly configured so that the external network (Layer4) cloud load balancer is provisioned automatically.
+For more information on additional cluster configuration, see [Cloud Provider Specifics](#cloud-provider-specifics).
+
+> **NOTE**: When external access is enabled, the NuoDB Helm Charts will create Internet-facing load balancers by default.  Be sure to understand the difference between _internet facing_ load balancers (allowing connectivity external to the cloud and Kubernetes cluster) and  _internal_ load balancers (allowing connectivity external to the Kubernetes cluster but within the cloud virtual network).
+
+If the SQL clients are located outside of the Kubernetes cluster where NuoDB is deployed but in the same cloud provider or virtual network, then _internal_ load balancer can be used.
+This is configured by setting the `admin.externalAccess.internalIP=true` and `database.te.externalAccess.internalIP=true` or further customized by explicitly setting custom annotations for the Kubernetes services using `admin.externalAccess.annotations` and `database.te.externalAccess.annotations` options.
```
Comment (jleslie85) [inline docs/HowToConnectExternally.md:47]: Suggest "using the"

**Review by jleslie85 — APPROVED** [review]
Excellent feature and document Stanimir!  I added a few minor wording suggestions. Thanks for adding this feature.

File: docs/HowToConnectExternally.md (hunk 3 above)
Comment (sivanov-nuodb) [reply docs/HowToConnectExternally.md:296]: Will remove the trailing "command" across the document.

File: docs/HowToConnectExternally.md [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+In earlier versions, it is only possible to connect to the provisioned TEs directly and bypass the NuoDB Admin.
+
+A Kubernetes service of type `LoadBalancer` or `NodePort` is created per TE group.
+This should allow the external SQL clients to connect to the TEs backing the service by uniquely targeting each TE group.
+Most of the cloud vendors provide Kubernetes Load Balancer controllers that support different service annotations used to control the properties and configuration of the provisioned cloud load balancer.
+The Kubernetes cluster should be properly configured so that the external network (Layer4) cloud load balancer is provisioned automatically.
+For more information on additional cluster configuration, see [Cloud Provider Specifics](#cloud-provider-specifics).
+
+> **NOTE**: When external access is enabled, the NuoDB Helm Charts will create Internet-facing load balancers by default.  Be sure to understand the difference between _internet facing_ load balancers (allowing connectivity external to the cloud and Kubernetes cluster) and  _internal_ load balancers (allowing connectivity external to the Kubernetes cluster but within the cloud virtual network).
```
Comment (adriansuarez) [inline docs/HowToConnectExternally.md:44]: What mechanism are you referring to when you say "internal load balancers"? "external to the Kubernetes cluster but within the cloud virtual network" sounds like it would be the same mechanism, e.g. `type=LoadBalancer`, but the DNS entry that is created is only resolvable within a private network.

File: docs/HowToConnectExternally.md [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+For more information, see [Use External Address](https://doc.nuodb.com/nuodb/latest/client-development/load-balancer-policies/#_use_external_address).
+
+Obtaining and configuring the address of the L4 load balancers can be tedious and error-prone as they are provisioned asynchronously by the Kubernetes controllers.
+NuoDB can inspect the Kubernetes services and configure the TE database processes with the `external-address` and `external-port` process labels automatically if the `--enable-external-access` process option is provided.
+This simplifies the deployment and ensures the correct configuration of the TE database processes.
+
+- For services of type `LoadBalancer`, NuoDB will configure the `external-address` process label with the service ingress IP or hostname as a value.
+- For services of type `NodePort` are provisioned, the customer must configure the  `external-address` and NuoDB will configure the `external-port` process label with the service node port as a value.
+If the `--enable-external-access` process option is supplied but the `external-address` process label is no defined, then external access won’t be enabled and a warning message will be logged.
```
Comment (adriansuarez) [inline docs/HowToConnectExternally.md:66]: Should say "process label is not defined". Also, I would say "`--enable-external-access` argument" rather than "process option", because this could be conflated with process options like `mem`.

File: docs/HowToConnectExternally.md [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+> **NOTE**: The node security configuration may also need to explicitly allow the NuoDB TE port (48006).
+
+### GCP
+
+No additional configuration is needed to enable NuoDB database external access in GKE.
+For more information, check [Configuring TCP/UDP load balancing](https://cloud.google.com/kubernetes-engine/docs/how-to/service-parameters).
+
+### AWS
+
+By default, the NuoDB Helm Charts 3.4.0 excepts that the [AWS Load Balancer Controller](https://docs.aws.amazon.com/eks/latest/userguide/aws-load-balancer-controller.html) is deployed in the Amazon Elastic Kubernetes Service (EKS) and the native CNI plugin is used when provisioning internet-facing load balancer.
```
Comment (adriansuarez) [inline docs/HowToConnectExternally.md:221]: Should say "expects" instead of "excepts".

File: docs/HowToConnectExternally.md [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+- incorrect NLB configuration
+- network connectivity problems including lack of routing, firewall configuration, and many more
+
+> **ACTION**: You can rule out any of the points above one by one.
+Start by checking the overall Pod status for the NuoDB domain and database.
+Some of the common troubleshooting steps are listed below, however, there might be additional verifications specific to your deployment.
+
+1. Verify that all AP, TE, and SM pods are reported _Ready_ using `kubectl get pods`.
+2. Check the NuoDB domain and database using `nuocmd show domain`.
+3. Verify the database availability inside the cluster is using the same connection properties as the application uses.
```
Comment (adriansuarez) [inline docs/HowToConnectExternally.md:272]: Should this say something like "Verify that the database is available to SQL clients within the Kubernetes cluster by using the same connection properties that the application uses"?

File: docs/HowToConnectExternally.md [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+To verify this, use the  `nuosql` tool inside an AP Pod.
+4. Make sure that `external-address` and/or `external-port` process labels are configured correctly on the TE database processes using `nuocmd --show-json-fields hostname,labels get processes`.
+If you are using the `--enable-external-access` process option, verify that the configured values are the same as the `EXTERNAL-IP` shown for the Kubernetes service in `kubectl get services` output.
+If the value is not the same, restart the TE database process and verify again.
+5. Verify that the `LBQuery` or `LBPolicy` syntax is correct. If using `LBPolicy`, verify that the expected policies are configured in the NuoDB Admin using `nuocmd get load-balancers` and `nuocmd get load-balancer-config`.
```
Comment (adriansuarez) [inline docs/HowToConnectExternally.md:277]: Another debugging tool that can be used to test `LBQuery` syntax is the `nuocmd get sql-connection` subcommand. I don't remember if a bogus `LBQuery` would actually generate a clear error message, or if the user would have to dig through `nuoadmin.log `(`nuoadmin` stdout).

File: docs/HowToConnectExternally.md (hunk 7 above)
Comment (adriansuarez) [reply docs/HowToConnectExternally.md:44]: So it looks like the mechanism here is using `internalIP`, which seems to be available in AWS by specifying certain annotations. Is it available in the other cloud providers, and if not should we make a note about it?

Comment (sivanov-nuodb) [reply docs/HowToConnectExternally.md:44]: This note here is specific to the load balancers provisioned in the cloud. The `database.te.externalAccess.internalIP` and `admin.externalAccess.internalIP` works in all supported cloud providers.

File: docs/HowToConnectExternally.md (hunk 8 above)
Comment (sivanov-nuodb) [reply docs/HowToConnectExternally.md:66]: Good catch! I will update this across the doc.

File: docs/HowToConnectExternally.md (hunk 9 above)
Comment (sivanov-nuodb) [reply docs/HowToConnectExternally.md:221]: Done

File: docs/HowToConnectExternally.md (hunk 10 above)
Comment (sivanov-nuodb) [reply docs/HowToConnectExternally.md:272]: Your suggestion read better. Will change the text.

File: docs/HowToConnectExternally.md (hunk 11 above)
Comment (sivanov-nuodb) [reply docs/HowToConnectExternally.md:277]: Good point!
I think it good enough for testing the LBQuery syntax:

```
bash-4.4$ nuocmd get sql-connection --db-name test --connection-properties LBQuery "random(labels(dummy foo))"
'get sql-connection' failed: Unable to get a SQL connection for database test: Next token does not match " *(\bcolocated\b|\baddress\b|\bor\b|\bstart_id\b|\bpid\b|\blabel\b|\bserver_id\b|\bany\b|\bhostname\b|\bnot\b|\band\b|\bregion\b|\bfirst\b|\bnode_id\b|\boption\b) *": random(|labels(dummy foo))

bash-4.4$ nuocmd get sql-connection --db-name test --connection-properties LBQuery "random(label(dummy foo))"

bash-4.4$ nuocmd get sql-connection --db-name test --connection-properties LBQuery "random(label(te te1))"
Process(address=test-te-1:48006, ...
```
